pulp_ansible/app/downloaders.py

```diff
@@ -0,0 +1,77 @@
+from logging import getLogger
+import asyncio
+import backoff
+import json
+
+from aiohttp.client_exceptions import ClientResponseError
+
+from pulpcore.plugin.download import http_giveup, HttpDownloader
+
+
+log = getLogger(__name__)
+
+
+class TokenAuthHttpDownloader(HttpDownloader):
+    """
+    Custom Downloader that automatically handles Token Based and Basic Authentication.
+
+    """
+
+    token_lock = asyncio.Lock()
+
+    def __init__(self, *args, **kwargs):
+        """
+        Initialize the downloader.
+        """
+        self.remote = kwargs.pop("remote")
+        super().__init__(*args, **kwargs)
+
+    @backoff.on_exception(backoff.expo, ClientResponseError, max_tries=10, giveup=http_giveup)
+    async def _run(self, extra_data=None):
+        """
+        Download, validate, and compute digests on the `url`. This is a coroutine.
+
+        This method is decorated with a backoff-and-retry behavior to retry HTTP 429 errors. It
+        retries with exponential backoff 10 times before allowing a final exception to be raised.
+
+        This method provides the same return object type and documented in
+        :meth:`~pulpcore.plugin.download.BaseDownloader._run`.
+
+        """
+        if not self.remote.token:
+            return await super()._run(extra_data=extra_data)
+
+        if not self.remote.auth_url:
+            headers = {"Authorization": self.remote.token}
+        else:
+            token = await self.update_token_from_auth_url()
+            headers = {"Authorization": "Bearer {token}".format(token=token)}
```

Comment:
Without `auth_url` will be considered as `community / galaxy.ansible.com / maybe standalone galaxy_ng`
With `auth_url` will be considered as `cloud.redhat.com`

Reply:
I think this is simpler than splitting the remotes. The only reservation I have is if users will be confused about this. I guess we could make this clear in the serializer help text.

Reply:
I'm just seeing this comment now, maybe we should leave it as one remote.

Reply:
pulp_container does something similar:
https://github.com/pulp/pulp_container/blob/master/pulp_container/app/downloaders.py#L25